src/compile.rs

```diff
@@ -329,6 +345,158 @@ pub fn build_func_map<'a>(
     Ok(decl_map)
 }
 
+pub fn validate_fs_context_duplicates(
```

Comment:
This function is long and deeply indented and therefore hard to follow.  Would it be possible to genericize the error cases a bit into a function or macro to make it more readable?  I see three instances of the "if let Some(unwrapped_error) = error" block with various differences that could perhaps be abstracted to make the whole thing more concise and readable.  I haven't read it all that closely yet though, so perhaps there's some challenges to that that make it not worth it?